x-pack/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/upgrade_review_algorithms.md

```diff
@@ -0,0 +1,245 @@
+And <field_name> field is updated by Elastic in this upgrade (target version != base version)
+And customized <field_name> field is the same as the Elastic update in this upgrade (current version == target version)
+Then for <field_name> field the diff algorithm should output the current version as the merged one without a conflict
+And <field_name> field should be returned from the `upgrade/_review` API endpoint
+And <field_name> field should be shown in the upgrade preview UI
+
+Examples:
+| algorithm          | field_name | base_version            | current_version         | target_version          | merged_version          |
+| single line string | name       | "A"                     | "B"                     | "B"                     | "B"                     |
+| number             | risk_score | 1                       | 2                       | 2                       | 2                       |
+| array of scalars   | tags       | ["one", "two", "three"] | ["one", "two", "four"]  | ["one", "two", "four"]  | ["one", "two", "four"]  |
+```
+
+### Rule field has an update and a custom value that are NOT the same - `ABC`
+
+#### **Scenario: `ABC` - Rule field is a number or single line string**
+
+**Automation**: 2 integration tests with mock rules + a set of unit tests for the algorithms
+
+```Gherkin
+Given <field_name> field is customized by the user (current version != base version)
+And <field_name> field is updated by Elastic in this upgrade (target version != base version)
+And customized <field_name> field is different than the Elastic update in this upgrade (current version != target version)
+Then for <field_name> field the diff algorithm should output the current version as the merged one with a non-solvable conflict
+And <field_name> field should be returned from the `upgrade/_review` API endpoint
+And <field_name> field should be shown in the upgrade preview UI
+
+Examples:
+| algorithm          | field_name | base_version | current_version | target_version  | merged_version |
+| single line string | name       | "A"          | "B"             | "C"             | "B"            |
+| number             | risk_score | 1            | 2               | 3               | 2              |
+```
+
+#### **Scenario: `ABC` - Rule field is an array of scalar values**
+
+**Automation**: 5 integration tests with mock rules + a set of unit tests for the algorithm
+
+```Gherkin
+Given <field_name> field is customized by the user (current version != base version)
+And <field_name> field is updated by Elastic in this upgrade (target version != base version)
+And customized <field_name> field is different than the Elastic update in this upgrade (current version != target version)
+Then for <field_name> field the diff algorithm should output a custom merged version with a solvable conflict
+And arrays should be deduplicated before comparison
+And arrays should be compared insensitive of case
```

Comment:
We ended up changing the logic for this, as explained in [this thread.](https://github.com/elastic/kibana/pull/186323#discussion_r1658805730) The diff should be case sensitive.

I would replace the last example for one that uses index patterns. In that case, it makes much more clear why it should be case-sensitive, as explained in the threat I linked above. We could have something like:

```
| algorithm        | field_name | base_version            | current_version         | target_version            | merged_version                 |
| array of scalars | index       | ["logs-*", "endgame-*", "endpoint-*"]   | ["Logs-*", "endgame-*"]          | ["logs-*", "endgame-*", "new-*"]         | ["Logs-*", "endgame-*", "new-*"]               |
| array of scalars | index       | ["logs-*"]   | ["logs-*", "Logs-*"]          | ["logs-*", "new-*"]         | ["logs-*", "Logs-*", "new-*"]               |
```

In the first case, "Logs-" would be preserved since it was added by the user, and "logs-" was removed by the user, so it does not appear in the final version.

In the second example, both are preserved, since they do not represent the same value to the algorithm.

Reply:
I can put it in the example, but `index` won't technically be covered by this algorithm as it's under the grouped data source field